### ISSUE TITLE
Possibility of support coroutine tests

### ISSUE THREAD

**Issue by kxepal** [DRMacIver/hypothesis#52]
I have a problem to test some stuff which based on asyncio. Default unittest doesn't likes when `test_` methods are coroutines - it has no idea what to do with them. Common workaround is to make coroutine as nested function like this:

```python
import asyncio
import aiohttp
from unittest import TestCase
from hypothesis import given

class DummyTestCase(TestCase):

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)

    def tearDown(self):
        self.loop.close()

    @given(str)
    def test_foo(self, x):
        @asyncio.coroutine
        def go():
            # dummy test just to check that hypothesis works with asyncio
            r = yield from aiohttp.request(x, 'http://localhost')
            assert r.status == 200, 'yay!'
        self.loop.run_until_complete(go())
```

The problem is that such solution is too much verbose and when you have thousands of tests it turns into a problem. However it works with hypothesis fine and this test fails as expected. To resolve inner coroutine problem I used the following trick:

```python
import asyncio
import aiohttp
import functools
from unittest import TestCase
from hypothesis import given


def run_in_loop(f):
    @functools.wraps(f)
    def wrapper(testcase, *args, **kwargs):
        coro = asyncio.coroutine(f)
        future = asyncio.wait_for(coro(testcase, *args, **kwargs),
                                  timeout=testcase.timeout)
        return testcase.loop.run_until_complete(future)
    return wrapper


class MetaAioTestCase(type):

    def __new__(cls, name, bases, attrs):
        for key, obj in attrs.items():
            if key.startswith('test_'):
                attrs[key] = run_in_loop(obj)
        return super().__new__(cls, name, bases, attrs)


class DummyTestCase(TestCase, metaclass=MetaAioTestCase):

    timeout = 5

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)

    def tearDown(self):
        self.loop.close()

    @given(str)
    def test_foo(self, x):
        # here we expect the fail, but it never happens
        r = yield from aiohttp.request(x, 'http://localhost')
        assert r.status == 200, 'yay!'
```

in other words by autodecorating `test_` methods to turn them into coroutines and run them within the event loop until they are completed. This removes a lot of routine and useless code, but suddenly such approach stops working with hypothesis - test passes as if it was never run.

I just found you project today and still playing around with it features without digging into `given` implementation, but may you have quick idea why the last case doesn't works?

**Comment by DRMacIver** [comment]
Eek. I'd not thought about this use case at all.

So I know why this doesn't work, and for the moment it's not going to. I'll have a think about how to support this as an explicit use case, probably by special casing functions which return generators, but I don't currently have any great ideas.

In order to understand what's happening:

When you pass a function to the @given decorator, Hypothesis is basically returning a new function which mimics the signature of the old, but what actually happens when called is that it is calling your function many times with a different arguments and seeing if any of them cause it to error. It doesn't pay any attention to the return value because it assume it's a "normal" unit test which just either returns None or errors, so it doesn't notice that you've given it this generator and expected it to do something about that, and as a result it just goes "Well I guess that worked then!" and tries the next value immediately.

**Comment by kxepal** [comment]
Aha, I see. Thanks for explanation. And because of that replacement I cannot run the target `test_X` coroutine in event loop since it's actually placed deeps inside of the `run_test_with_generator` wrapper. What my metaclass tries to do here is to run that wrapper as coroutine what's certainly makes no sense. I think I know how to fix that. Will propose it shortly.

**Comment by DRMacIver** [comment]
If you had patching Hypothesis to make this work in mind I'd recommend against it for now. The runner loop is some of the oldest and hairiest code in the codebase and has been patched rather than refactored about 10 too many times. Sorting it out so it's less of a nightmare is on my TODO list and it will probably happen in the not too distant future, but I think trying to patch this in right now will just make it worse.

Could something like the following work for you instead?

```python
@given(str)
@runinloop
def test_foo(self, x):
        # here we expect the fail, but it never happens
        r = yield from aiohttp.request(x, 'http://localhost')
        assert r.status == 200, 'yay!'```

This will move your runinloop call to inside the bit that Hypothesis has access to, so it can invoke it as if it were just a normal function.

**Comment by kxepal** [comment]
Oh, sorry - I made a PR right after you commented. Yes, this may work and that's technically what I did in #53, just on hypothesis side to handle coroutines in friendly way. Think that was bad decision?

**Comment by DRMacIver** [comment]
Yeah, sorry, I've left a longer comment on your PR but basically I'm up for making this possible but not really up for the solution you proposed. I'll try to make sure something like that is straightforward soon though.

**Comment by DRMacIver** [comment]
(As a side note, you'd also have had to [sign the CLA](http://hypothesis.readthedocs.org/en/latest/contributing.html) before I'd have been able to accept it)

**Comment by DRMacIver** [comment]
This should now be possible to do using the new executors feature: http://hypothesis.readthedocs.org/en/master/details.html#custom-function-execution

I'm probably not going to support asyncio explicitly for now, but you should be able to just hook into this feature.

Let me know if this does/doesn't work

**Comment by kxepal** [comment]
Thanks for the update! Going try it now.

**Comment by kxepal** [comment]
So far it was great. `execute_example` method takes `<function reify_and_execute.<locals>.run at ...>` as input argument which on call may return coroutine to let me execute it within the loop. All great - tests which doesn't raises any errors passes fine and actually makes expected io. However, problem starts with the bad cases:

```
import asyncio
import aiohttp
from unittest import TestCase
from hypothesis import given


class DummyTestCase(TestCase):

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)

    def tearDown(self):
        self.loop.close()

    @given(str)
    @asyncio.coroutine
    def test_thing(self, thing):
        r = yield from aiohttp.request('GET', 'http://localhost/' + thing)
        assert r.status == 200

    def execute_example(self, f):
        self.loop.run_until_complete(f())
```


```
  File "/usr/lib64/python3.3/unittest/case.py", line 384, in _executeTestPart
    function()
  File "/home/kxepal/projects/sandbox/.hypothesis/eval_source/hypothesis_temporary_module_6b86962c6ced65e40a665c61f11a8384a5317a33_0.py", line 5, in test_foo
    return f(self=self, thing=thing)
  File "/usr/local/opt/virtualenv/sandbox/lib/python3.3/site-packages/hypothesis/core.py", line 423, in wrapped_test
    print_example=True
  File "./test.py", line 25, in execute_example
    self.loop.run_until_complete(f())
  File "/usr/local/opt/virtualenv/sandbox/lib/python3.3/site-packages/hypothesis/core.py", line 255, in run
    return test(*args, **kwargs)
  File "/usr/local/opt/virtualenv/sandbox/lib/python3.3/site-packages/hypothesis/core.py", line 214, in test_or_flaky
    raise Flaky(test, (args, kwargs))
hypothesis.errors.Flaky: Hypothesis 'test_thing' produces unreliable results: ((), {'thing': '', 'self': <test.DummyTestCase testMethod=test_thing>}) falsified it on the first call but did not on a subsequent one
```

AFAICS the problem is that `core.test_must_fail` activates coroutine within additional decorator where it cannot be executed by the `execute_example` within the loop what leaves no choice but to raise a Flaky exception instead.

**Comment by DRMacIver** [comment]
Bad. Yes. You are of course entirely right. I'll have a patch release out for you tomorrow which should fix this.

**Comment by kxepal** [comment]
Great! Staying tuned.

**Comment by DRMacIver** [comment]
OK. I've now done what a sensible person would do and actually tested your use case to make sure it works. :-) I have released 1.2.1, and the following code works (in the sense of "the test fails in the way it should") for me:

```python
import asyncio
import aiohttp
import functools
from unittest import TestCase
from hypothesis import given


class DummyTestCase(TestCase):

    timeout = 5

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)

    def tearDown(self):
        self.loop.close()

    def execute_example(self, f):
        coro = asyncio.coroutine(f)
        future = asyncio.wait_for(coro(),
                                  timeout=self.timeout)
        return self.loop.run_until_complete(future)

    @given(str)
    def test_foo(self, x):
        r = yield from aiohttp.request(x, 'http://localhost')
        assert r.status == 200, 'yay!'
````

**Comment by kxepal** [comment]
@DRMacIver Works like a charm! Thank you a lot!
